core/state/state_object.go

```diff
@@ -310,16 +310,23 @@ func (s *stateObject) setState(key, value common.Hash) {
 // finalise moves all dirty storage slots into the pending area to be hashed or
 // committed later. It is invoked at the end of every transaction.
 func (s *stateObject) finalise(prefetch bool) {
-	slotsToPrefetch := make([][]byte, 0, len(s.dirtyStorage))
-	for key, value := range s.dirtyStorage {
-		s.pendingStorage[key] = value
-		if value != s.originStorage[key] {
-			slotsToPrefetch = append(slotsToPrefetch, common.CopyBytes(key[:])) // Copy needed for closure
+	if prefetch && s.db.prefetcher != nil && s.data.Root != emptyRoot {
+		slotsToPrefetch := make([][]byte, 0, len(s.dirtyStorage))
+		for key, value := range s.dirtyStorage {
+			s.pendingStorage[key] = value
+			if value != s.originStorage[key] {
+				slotsToPrefetch = append(slotsToPrefetch, common.CopyBytes(key[:])) // Copy needed for closure
+			}
+		}
+		if len(slotsToPrefetch) > 0 {
+			s.db.prefetcher.prefetch(s.data.Root, slotsToPrefetch)
+		}
+	} else {
+		for key, value := range s.dirtyStorage {
+			s.pendingStorage[key] = value
```

Comment:
I slightly prefer moving this to the beginning of the func instead of adding an `else` clause

Reply:
Updated

Reply:
@s1na Weird, the above change causes build failure, I'm reverting it to see whether it will pass

Reply:
Restarting the build for that commit to make sure it wasn't because of a flaky test

Reply:
so it passed the second time around

Reply:
Cool, let me recover the code!

Reply:
lolling at the commit title `Revert "Revert "fix for review""` :)) sorry we went so much back and forth for a nitpick